### ISSUE TITLE
InvalidOperationException: Shutdown has already been called in C# .NET Core simple GRPC client/server

### ISSUE THREAD

**Issue by bobuva** [grpc/grpc#14823]
I've got a bi-directional streaming service implemented in C# using grpc with protobuf-generated C# classes. The service proto looks like this:

```
service AgentService {
  rpc OpenNotificationStream (stream NotificationMessage) returns (stream ServerMessage);
}
```

 My server code looks like this:

```    public class AgentServerImpl : AgentService.AgentServiceBase
    {
        public AgentServerImpl()  { }

        public override async Task OpenNotificationStream(
            IAsyncStreamReader<NotificationMessage> requestStream,
            IServerStreamWriter<ServerMessage> responseStream,
            ServerCallContext context)
        {
            while(await requestStream.MoveNext())
            {
                var notification = requestStream.Current;
                if (notification != null)
                {
                    Console.WriteLine("Received: {0}", notification.ToString());
                }

                ServerMessage serverMessage = new ServerMessage();
                await responseStream.WriteAsync(serverMessage);
            }
        }
    }
```
In the client, I do the following:

```
    // Main program 
    class Program
    {
        static void Main(string[] args)
        {
            var backgroundTask = Task.Run(() => DoConnectAsync());
            backgroundTask.Wait();
        }

        private static async void DoConnectAsync()
        {
            var proxyClient = new ProxyClient();
            await proxyClient.MakeRequest();
        }
    }

    // ProxyClient
    public class ProxyClient : IDisposable
    {
        private Channel channel;
        private AsyncDuplexStreamingCall<NotificationMessage, ServerMessage> notificationCall;
        private CancellationToken cancellationToken;
        private const string DefaultProxyHost = "localhost";
        private const int DefaultProxyPort = 50050;

        public ProxyClient()
        {
            this.cancellationToken = new CancellationToken(false);
            initializeInsecureGrpc();
        }

        private void initializeInsecureGrpc()
        {
            try
            {
                channel = new Channel(DefaultProxyHost, DefaultProxyPort, ChannelCredentials.Insecure);
                var agentService = new AgentService.AgentServiceClient(channel);
                notificationCall = agentService.OpenNotificationStream();
            }
            catch (Exception e)
            {
                Logger.Log("Exception " + e.Message);
            }

        }

        public async Task MakeRequest()
        {
           // This just returns a newly-allocated request message. Code not included in this sample.
            var sprRequestMessage = createSprRequestMessage();
            
            await notificationCall.RequestStream.WriteAsync(sprRequestMessage);
            
            await notificationCall.RequestStream.CompleteAsync();

            // Check for ServerMessage responses
            await handleProxyResponse();

            await channel.ShutdownAsync();
        }

        private async Task handleProxyResponse()
        {
            var responseReaderTask = Task.Run(async () =>
            {
                if (notificationCall == null)
                {
                    Console.WriteLine("handleProxyResponse: notificationCall is NULL");
                }
                else
                {
                    Console.WriteLine("handleProxyResponse: notificationCall = {0}", notificationCall.ToString());

                    while (await notificationCall.ResponseStream.MoveNext(new CancellationToken(false)))
                    {
                        var serverMsg = notificationCall.ResponseStream.Current;
                        if (serverMsg.Err != null)
                        {
                            Logger.Log(serverMsg.Err.Msg);
                        }

                        Logger.Log("Got message \"{0}\"", serverMsg.ToString());
                    }
                }
            });

            await responseReaderTask;
        }

       // IDisposable pattern implementation not included for brevity.

```

I'm using:
    grpc 1.10.0
    Google.Protobuf 3.5.1
    .NET Core 2.1
    Windows 10
     Microsoft Visual Studio compiler

Steps I've taken:

1. Start the server
2. Start the client

Client throws an exception:

```
initializeInsecureGrpc: notificationCall = Grpc.Core.AsyncDuplexStreamingCall`2[io.shiftleft.proto.agent.NotificationMessage,io.shiftleft.proto.agent.ServerMessage]
handleProxyResponse: notificationCall = Grpc.Core.AsyncDuplexStreamingCall`2[io.shiftleft.proto.agent.NotificationMessage,io.shiftleft.proto.agent.ServerMessage]
Got message "{ }"

Unhandled Exception: System.InvalidOperationException: Shutdown has already been called
   at Grpc.Core.Internal.CompletionQueueSafeHandle.BeginOp() in c:\jenkins\workspace\gRPC_build_packages\platform\windows\src\csharp\Grpc.Core\Internal\CompletionQueueSafeHandle.cs:line 108
   at Grpc.Core.Internal.CallSafeHandle.StartReceiveMessage(IReceivedMessageCallback callback) in c:\jenkins\workspace\gRPC_build_packages\platform\windows\src\csharp\Grpc.Core\Internal\CallSafeHandle.cs:line 145
   at Grpc.Core.Internal.AsyncCallBase`2.ReadMessageInternalAsync() in c:\jenkins\workspace\gRPC_build_packages\platform\windows\src\csharp\Grpc.Core\Internal\AsyncCallBase.cs:line 158
   at Grpc.Core.Internal.ClientResponseStream`2.<MoveNext>d__5.MoveNext() in c:\jenkins\workspace\gRPC_build_packages\platform\windows\src\csharp\Grpc.Core\Internal\ClientResponseStream.cs:line 55
   at ShiftLeft.Controller.ProxyClient.<<handleProxyResponse>b__18_0>d.MoveNext() in C:\src\ShiftLeftSecurity\sl-microagent-dotnet\sl-dotnetAgent\ShiftLeft.Controller\ProxyClient.cs:line 170
   at ShiftLeft.Controller.ProxyClient.<handleProxyResponse>d__18.MoveNext() in C:\src\ShiftLeftSecurity\sl-microagent-dotnet\sl-dotnetAgent\ShiftLeft.Controller\ProxyClient.cs:line 183
   at ShiftLeft.Controller.ProxyClient.<RequestSpr>d__15.MoveNext() in C:\src\ShiftLeftSecurity\sl-microagent-dotnet\sl-dotnetAgent\ShiftLeft.Controller\ProxyClient.cs:line 119
   at ShiftLeft.Controller.Program.<DoConnectAsync>d__2.MoveNext() in C:\src\ShiftLeftSecurity\sl-microagent-dotnet\sl-dotnetAgent\ShiftLeft.Controller\Program.cs:line 40
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
Press any key to continue . . .
```
 As you can see above, the `handleProxyResponse` method which waits for a response from the server, does get one, but it is empty. Sometimes, like when I try to send more than one message to the server, I see some of them getting to the server, but no response in the client.

In the server console output, I see the expected request being received:

```
AgentService server listening on port 50050
Press any key to stop the server...
Received: { "request": { "clientId": "EWuDLV064fuUvJcQYXQRHijcnzo", "sprId": "sl / 48383d99-02db-42c8-b703-01934cbd44ca / cli - dotnet-agent-controller - 0.0.1.exe", "timestamp": "131666387947394865", "processors": "2", "osName": "Win", "osArch": "x64", "osVersion": "10", "license": "MS4wLjAuMA==", "nonStrict": true, "sprState": "SPR_NOT_USED" } }
```

I don't know why I'm getting the `InvalidOperationException` with the `Shutdown has already been called` message.

**Comment by bobuva** [comment]
Closing this issue as I've figured out that my problem was not with `grpc` but with how I wrote the asynchronous code using C#'s `async await` features.

**Comment by AlexTelon** [comment]
@bobuva Do you remember what the specific issue was? I happen to have the same issue and when googling "shutdown has already been called c#" this issue was the top result for me. So if you could take the time to expand on what your issue was it would help me and hopefully others greatly! :)

**Comment by bobuva** [comment]
@AlexTelon For sure, sorry I didn't do that at the time. I rewrote it so that

1. I don't call `RequestStream.CompleteAsync()` and `channel.ShutdownAsync()` right away. In fact I left it for when I might close the entire connection.

2. I moved the `await handleProxyResponse()` call to a separately-running task, kicked off at a higher level, rather than from within the `MakeRequest` method as shown in my original post. 

I completely rewrote my code, so it's possible there was something else, but I believe that was what took care of the `Shutdown has already been called` exception.
